File: pipeline/aeb/AEBPipeline.m
classdef AEBPipeline < handle
    % AEBPipeline: Orchestrates AEB data processing pipeline
    %
    % Chains Config, InputHandler, EventDetector, KPIExtractor, and Visualizer
    % to process MF4 files, detect AEB events, extract KPIs, and visualize results.
    
    properties
        Config        % Instance of Config class
        InputHandler  % Handles raw data loading
        EventDetector % Segments raw data into AEB events
        KPIExtractor  % Extracts KPIs from each event
        Visualizer    % Generates plots
    end
    
    methods
        function obj = AEBPipeline(configPath)
            % Constructor: Initialize pipeline components
            %
            % Inputs:
            %   configPath - Path to the Config JSON file
            
            if nargin < 1
                error('Configuration file path is required.');
            end
            validateattributes(configPath, {'char', 'string'}, {'nonempty'}, ...
                'AEBPipeline', 'configPath');
            
            % Step 1: Create Config instance
            try
                obj.Config = Config.fromJSON(configPath);
            catch err
                error('Failed to create Config instance: %s', err.message);
            end
            
            % Step 2: Create InputHandler instance
            try
                obj.InputHandler = InputHandler(obj.Config);
            catch err
                error('Failed to create InputHandler instance: %s', err.message);
            end
        end
        
        function run(obj)
            % Run the AEB processing pipeline
            %
            % Executes the pipeline steps: process MF4 files, detect AEB events,
            % extract KPIs, export to CSV, and visualize results.

            fprintf('\n🚀 Starting AEB processing pipeline...\n');

            % Step 3: Process MF4 files
            fprintf(' ➡️ [Step 1/4] Processing MF4 files...\n');
            try
                processedData = obj.InputHandler.processMF4Files();
                if isempty(processedData)
                    warning('No data processed from MF4 files. Aborting pipeline.');
                    return;
                end
                fprintf('    ✅ MF4 files processed successfully.\n');
            catch err
                error('  ❌ Failed to process MF4 files: %s', err.message);
            end

            % Step 4: Detect AEB events
            fprintf(' ➡️ [Step 2/4] Detecting AEB events...\n');
            try
                obj.EventDetector = EventDetector(obj.InputHandler);
                obj.EventDetector.processAllFiles();
                fprintf('    ✅ AEB events detected successfully.\n');
            catch err
                error('  ❌ Failed to process AEB events: %s', err.message);
            end

            % Step 5: Extract KPIs
            fprintf(' ➡️ [Step 3/4] Extracting KPIs and exporting to CSV...\n');
            try
                obj.KPIExtractor = KPIExtractor(obj.Config, obj.EventDetector);
                obj.KPIExtractor = obj.KPIExtractor.processAllMatFiles();
                obj.KPIExtractor.exportToCSV();
                fprintf('    ✅ KPIs extracted and exported to CSV successfully.\n');
            catch err
                error('  ❌ Failed to extract KPIs or export to CSV: %s', err.message);
            end

            % Step 6: Visualize results
            fprintf(' ➡️ [Step 4/4] Generating visualizations...\n');
            try
                obj.Visualizer = Visualizer(obj.Config, obj.KPIExtractor);
                obj.Visualizer.plot();
                fprintf('    ✅ Visualizations generated successfully.\n');
            catch err
                error('  ❌ Failed to generate visualizations: %s', err.message);
            end

            fprintf('\n🎉 AEB processing pipeline completed successfully.\n');
        end % function run

    end % methods
end

File: pipeline/aeb/aeb_pipeline.py
import os
import warnings
import pandas as pd
from pathlib import Path

from config.config import Config
from pipeline.aeb.aeb_event_segmenter import AebEventSegmenter
from pipeline.aeb.aeb_kpi_extractor import AebKpiExtractor
from pipeline.aeb.visualizer import AebVisualizer


class AebPipeline:
    """
    AebPipeline
    ------------
    High-level orchestrator for the AEB KPI processing pipeline.

    Steps:
      1. Load configuration (JSON)
      2. Process MF4 files (resample + signal extraction)
      3. Detect AEB events
      4. Extract KPIs and export to Excel
      5. Visualize results (interactive or static)
    """

    def __init__(self, config_path: Path, input_handler=None):
        self.config_path = Path(config_path)
        self.cfg         = None
        self.ih          = input_handler   # can be injected externally
        self.event       = None
        self.kpi         = None
        self.viz         = None

    # ------------------------------------------------------------------ #
    def run(self, skip_mf4_processing: bool = False):
        """Run the entire AEB pipeline sequentially."""
        print("\n🚀 Starting AEB Processing Pipeline...\n")

        # 1️⃣ Load configuration
        self._load_config()

        # 2️⃣ Process MF4 files (skip if external handler already provided)
        if self.ih is None and not skip_mf4_processing:
            self._process_mf4_files()
        else:
            print("🪄 Using external InputHandler instance.")

        # 3️⃣ Run event detection
        self._detect_events()

        # 4️⃣ Extract KPIs
        self._extract_kpis()

        # 5️⃣ Visualize results
        self._visualize_results()

        print("\n🎉 AEB Pipeline finished successfully.\n")

    # ------------------------------------------------------------------ #
    def _load_config(self):
        """Step 1: Load Config JSON."""
        print("➡️ [1/5] Loading configuration...")
        if not self.config_path.exists():
            raise FileNotFoundError(f"⚠️ Config file not found: {self.config_path}")
        self.cfg = Config.from_json(self.config_path)
        print(f"✅ Config loaded: {self.config_path}")

    # ------------------------------------------------------------------ #
    def _process_mf4_files(self):
        """Step 2: Process MF4 files using InputHandler."""
        print("\n➡️ [2/5] Processing MF4 files...")
        from pipeline.input_handler import InputHandler  # imported only when needed
        try:
            self.ih = InputHandler(self.cfg)
            self.ih.process_mf4_files()
            print("✅ MF4 files processed successfully.")
        except Exception as e:
            raise RuntimeError(f"❌ MF4 processing failed: {e}")

    # ------------------------------------------------------------------ #
    def _detect_events(self):
        """Step 3: Detect AEB events."""
        print("\n➡️ [3/5] Detecting AEB events...")
        try:
            self.event = AebEventSegmenter(self.ih, self.cfg)
            print("🚦 Running event detection...\n")
            self.event.process_all_files()
            print("✅ Event detection finished.\n")
        except Exception as e:
            raise RuntimeError(f"❌ Event detection failed: {e}")

    # ------------------------------------------------------------------ #
    def _extract_kpis(self):
        """Step 4: Extract KPIs and export to Excel."""
        print("\n➡️ [4/5] Extracting KPIs...")
        try:
            self.kpi = AebKpiExtractor(self.cfg, self.event)
            self.kpi.process_all_mdf_files()
            self.kpi.export_to_excel()
            print("✅ KPI extraction and Excel export done.")
        except Exception as e:
            raise RuntimeError(f"❌ KPI extraction failed: {e}")

    # ------------------------------------------------------------------ #
    def _visualize_results(self):
        """Step 5: Visualization."""
        print("\n➡️ [5/5] Launching visualization...\n")
        try:
            self.viz = AebVisualizer(self.cfg, self.kpi)
            self.viz.interactive = True  # enable zoomable interactive plots
            self.viz.plot()
        except Exception as e:
            raise RuntimeError(f"❌ Visualization failed: {e}")
